### DISCUSSION TITLE
How to integrate sensors with the "Pixel LEDS" sketch?

### DISCUSSION THREAD

**Discussion by FidelDworp** [HomeSpan/HomeSpan#568] [Ideas]
My [previous thread](https://github.com/HomeSpan/HomeSpan/discussions/558) can be closed successfully, thanks!

As I explained in that thread, I’m on my way to replace the 10 Particle Photon controllers in my Home automation system by a number of ESP32’s powered by HomeSpan, while still keeping my existing peripherals (PowerPiXels and Roomsense boxes) in use…

My PowerPixels (= NeopiXels on steroids!) now work perfectly already with the sketch you made in the previous thread.
Now I want to try the second step: Since about 10 years, I use a combination of 5 sensors in the form of my “RoomSense” boxes in every room of my home:

![RoomSenseBox-Kitchen](https://user-images.githubusercontent.com/46328609/235194300-bfd1894f-f801-4a58-80db-cc004fd9831d.jpg)

![RoomSenseBox-PCB](https://user-images.githubusercontent.com/46328609/235194341-20f02ea0-cc8c-4680-9b56-5468185d31a7.jpeg)

-  These are the five functions of my “RoomSense” boxes I want to integrate these in one ESP32:
    1. Individually controllable Neopixels (Is working!)
    2. Occupancy sensor (PIR or electrostatic “radar” sensor)
    3. Light sensor (LDR + resistor)
    4. DHT22 sensor (Temperature & Humidity)
    5. DS18B20 Precision temperature sensor

When an ESP32 works with these peripherals, I want to create the following “Automations” in Apple Homekit: 

    1. When an occupancy sensor responds to movement, lamp(s) are switched ON for some period. They should switch off when no occupancy is detected anymore.
    2. When the light sensor value exceeds a certain threshold, the occupancy sensor is disabled (during daytime) so that lamps won’t switch ON.

- As a first step, I have set-up a breadboard with following elements:
    1. Occupancy sensor: Pin 25 (pin-compatible with “HC-SR501” PIR sensor)
    2. Neopixel strip: Pin 26 (Works!)
    3. Light sensor: Pin 27

![ESP32_Test-set-up](https://user-images.githubusercontent.com/46328609/235194505-6a833117-2474-4525-a8f6-10e7a6e94f98.jpeg)

Note: I checked and found ESP32’s GPIO 25, 26 & 27 pins are 5V tolerant.

- I am now trying to make a sketch, (based on [your Pixel example](https://github.com/HomeSpan/HomeSpan/discussions/558)), which creates following objects in Apple Homekit:
    - 3 Lamps (= Working)
    - 1 Occupancy sensor
    - 1 Light sensor

- I have already studied following Homespan documentation:
    1. read the API text (it is above my “pay grade”…)
    2. Checked-out out all relevant tutorial sketches and I tried 5 out.

- I found a “DEV_MotionSensor” structure in [this thread](https://github.com/HomeSpan/HomeSpan/discussions/159), for the popular “HC-SR501” PIR sensor, which sounded close enough to what I am looking for:

- Then I tried to integrate that structure in your Pixel example sketch. Below sketch is what I have tested till now:

```cpp
// THIS SKETCH is an exercise combining A NEOPIXEL structure + a PIR sensor structure

#define NEOPIXEL_RGB_PIN       26           // PIN USED FOR NEOPIXEL STRAND
#define NEOPIXEL_SIZE          3            // TOTAL NUMBER OF PIXELS IN NEOPIXEL STRAND
 
#include "HomeSpan.h"
#include "extras/Pixel.h"                   // include the HomeSpan Pixel class

// CREATE GLOBAL OBJECTS TO STORE ALL COLORS IN A SINGLE STRAND
Pixel pixelStrand(NEOPIXEL_RGB_PIN);                          // create a single, global instance of pixelStrand
Pixel::Color colorList[NEOPIXEL_SIZE]={Pixel::RGB(0,0,0)};    // create an array to hold all the pixel colors - initialize to all zeros

///////////////////////////////

struct NeoPixel_RGB : Service::LightBulb      // Addressable single-wire NeoPixel Strand
{
  Characteristic::On power{0,true};
  Characteristic::Hue H{0,true};
  Characteristic::Saturation S{0,true};
  Characteristic::Brightness V{100,true};

  uint8_t startingColor;
  uint8_t nColors;
  
  NeoPixel_RGB(uint8_t startingPixel, uint8_t nPixels) : Service::LightBulb()
  {
    V.setRange(5,100,1);                      // sets the range of the Brightness to be from a min of 5%, to a max of 100%, in steps of 1%
    this->startingColor=startingPixel-1;      // start of color array for this subset of pixels
    this->nColors=nPixels;                    // number of colors to change (which equals number of pixels in this subset)
    update();                                 // manually call update() to set pixel with restored initial values
  }

  boolean update() override
  {
    int p=power.getNewVal();
    
    float h=H.getNewVal<float>();       // range = [0,360]
    float s=S.getNewVal<float>();       // range = [0,100]
    float v=V.getNewVal<float>();       // range = [0,100]

    for(int i=0;i<nColors;i++)                  // loop over a SUBSET of pixels and change their color
      colorList[i+startingColor]=Pixel::HSV(h*p, s*p, v*p);

    pixelStrand.set(colorList,NEOPIXEL_SIZE);   // refresh ALL pixels in the strand based on the updated colors
          
    return(true);  
  }
};

struct DEV_MotionSensor : Service::MotionSensor  // Motion sensor
{
  SpanCharacteristic *movement;                  // reference to the MotionDetected Characteristic
  int sensorPin = 25;                            // pin number of the sensor

  DEV_MotionSensor(int sensorPin) : Service::MotionSensor()
  {
    
    this->sensorPin = sensorPin;
    pinMode(sensorPin,INPUT);
    boolean motion = digitalRead(sensorPin);
    movement=new Characteristic::MotionDetected(motion);// instantiate the MotionDetected Characteristic

  }                                                     // end constructor

  void loop()
  {
    boolean motion = digitalRead(sensorPin);
    if (motion != movement->getVal())
    {
      movement->setVal(motion);
      if (motion == true)
      {
        char c[64];
        sprintf(c,"Motion was detected\n");
        LOG1(c);          
      }
    }
  }
};

///////////////////////////////

void setup()
{
  
  Serial.begin(115200);
 
  homeSpan.begin(Category::Lighting,"Pixel LEDS"); // Standard pairing code: 466-37-726 => Use standard scannable code!


  SPAN_ACCESSORY();             // create Bridge (note this sketch uses the SPAN_ACCESSORY() macro, introduced in v1.5.1 --- see the HomeSpan API Reference for details on this convenience macro)
  
  SPAN_ACCESSORY("Light 1");    // Here we use the SPAN_ACCESSORY() macro and specify an optional name for the first set of LEDs
    new NeoPixel_RGB(1,1);      // create a LightBulb that controls pixel 1 of a Neo Pixel Strand (we start at pixel 1 with length of 1 pixel)

  SPAN_ACCESSORY("Light 2");    // Repeat for second set of LEDs
    new NeoPixel_RGB(2,1);      // create a LightBulb that controls pixel 2 of a Neo Pixel Strand (we start at pixel 2 with length of 1 pixel)

  SPAN_ACCESSORY("Light 3");    // Repeat for third set of LEDs
    new NeoPixel_RGB(3,1);      // create a LightBulb that controls pixel 3 of a Neo Pixel Strand (we start at pixel 3 with length of 1 pixel)
}

///////////////////////////////

void loop()
{
  homeSpan.poll();
}

///////////////////////////////
```

- **Result**: it compiles OK. After uploading this sketch, I paired the ESP32 successfully with Homekit. However, only the three Pixel lights appear… (I’m already glad this works!)

- **Question**: What do I have to do to also make the **Occupancy sensor** and **Light sensor** appear in Homekit?

Thanks on beforehand!
FiDelDworp

**Comment by HomeSpan** [comment]
In the above sketch you added in the *definition* a motion sensor (this is section that starts with `struct DEV_MotionSensor...` but you did not use it anywhere (i.e. it's only a definition of something that can be used).   To create a DEV_MotionSensor object you need to call it with `new` somewhere in your sketch.   In HomeSpan, the way to do this would be as a new Accessory added to the three already there (Light 1, Light 2, and Light 3).  You could do this by adding the following:

```C++
SPAN_ACCESSORY("Motion Sensor");
  new DEV_MotionSensor(motionSensorPin);
```

When HomeSpan first starts it will print out all the Accessories it finds - that's how you'll know you implemented correctly.

Note when you add/delete Accessories in a sketch, the Home App MAY correctly recognize it, but it also may get confused - Apple is not very consistent around this.  If the sketch compiles and runs correctly, but the new Accessory does not show up in the Home App, first try closing and restarting the Home App.  If that does not solve the problem you'll need to unpair the Accessory from the Home App, type 'H' into the Serial Monitor to cause HomeSpan to generate a new Device ID, and the re-pair the Accessory with the Home App again.

**Reply by FidelDworp** [reply]
Thanks for the tips!
I'm trying to follow but as I said: It's above my pay-grade... I'm not en educated programmer: Only started when retiring...

I integrated the new accessory from your previous message in the sketch as follows:

```cpp
// THIS SKETCH is an exercise combining A NEOPIXEL structure + a PIR sensor structure

#define NEOPIXEL_RGB_PIN       26           // PIN USED FOR NEOPIXEL STRAND
#define NEOPIXEL_SIZE          3            // TOTAL NUMBER OF PIXELS IN NEOPIXEL STRAND
 
#include "HomeSpan.h"
#include "extras/Pixel.h"                   // include the HomeSpan Pixel class

// CREATE GLOBAL OBJECTS TO STORE ALL COLORS IN A SINGLE STRAND
Pixel pixelStrand(NEOPIXEL_RGB_PIN);                          // create a single, global instance of pixelStrand
Pixel::Color colorList[NEOPIXEL_SIZE]={Pixel::RGB(0,0,0)};    // create an array to hold all the pixel colors - initialize to all zeros

///////////////////////////////

struct NeoPixel_RGB : Service::LightBulb      // Addressable single-wire NeoPixel Strand
{
  Characteristic::On power{0,true};
  Characteristic::Hue H{0,true};
  Characteristic::Saturation S{0,true};
  Characteristic::Brightness V{100,true};

  uint8_t startingColor;
  uint8_t nColors;
  
  NeoPixel_RGB(uint8_t startingPixel, uint8_t nPixels) : Service::LightBulb()
  {
    V.setRange(5,100,1);                      // sets the range of the Brightness to be from a min of 5%, to a max of 100%, in steps of 1%
    this->startingColor=startingPixel-1;      // start of color array for this subset of pixels
    this->nColors=nPixels;                    // number of colors to change (which equals number of pixels in this subset)
    update();                                 // manually call update() to set pixel with restored initial values
  }

  boolean update() override
  {
    int p=power.getNewVal();
    
    float h=H.getNewVal<float>();       // range = [0,360]
    float s=S.getNewVal<float>();       // range = [0,100]
    float v=V.getNewVal<float>();       // range = [0,100]

    for(int i=0;i<nColors;i++)                  // loop over a SUBSET of pixels and change their color
      colorList[i+startingColor]=Pixel::HSV(h*p, s*p, v*p);

    pixelStrand.set(colorList,NEOPIXEL_SIZE);   // refresh ALL pixels in the strand based on the updated colors
          
    return(true);  
  }
};

struct DEV_MotionSensor : Service::MotionSensor  // Motion sensor
{
  SpanCharacteristic *movement;                  // reference to the MotionDetected Characteristic
  int sensorPin = 25;                            // pin number of the sensor

  DEV_MotionSensor(int sensorPin) : Service::MotionSensor()
  {
    
    this->sensorPin = sensorPin;
    pinMode(sensorPin,INPUT);
    boolean motion = digitalRead(sensorPin);
    movement=new Characteristic::MotionDetected(motion);// instantiate the MotionDetected Characteristic

  }                                                     // end constructor

  void loop()
  {
    boolean motion = digitalRead(sensorPin);
    if (motion != movement->getVal())
    {
      movement->setVal(motion);
      if (motion == true)
      {
        char c[64];
        sprintf(c,"Motion was detected\n");
        LOG1(c);          
      }
    }
  }
};

///////////////////////////////

void setup()
{
  
  Serial.begin(115200);
 
  homeSpan.begin(Category::Lighting,"Pixel LEDS"); // Standard pairing code: 466-37-726 => Use standard scannable code!


  SPAN_ACCESSORY();             // create Bridge (note this sketch uses the SPAN_ACCESSORY() macro, introduced in v1.5.1 --- see the HomeSpan API Reference for details on this convenience macro)
  
  SPAN_ACCESSORY("Light 1");    // Here we use the SPAN_ACCESSORY() macro and specify an optional name for the first set of LEDs
    new NeoPixel_RGB(1,1);      // create a LightBulb that controls pixel 1 of a Neo Pixel Strand (we start at pixel 1 with length of 1 pixel)

  SPAN_ACCESSORY("Light 2");    // Repeat for second set of LEDs
    new NeoPixel_RGB(2,1);      // create a LightBulb that controls pixel 2 of a Neo Pixel Strand (we start at pixel 2 with length of 1 pixel)

  SPAN_ACCESSORY("Light 3");    // Repeat for third set of LEDs
    new NeoPixel_RGB(3,1);      // create a LightBulb that controls pixel 3 of a Neo Pixel Strand (we start at pixel 3 with length of 1 pixel)

  SPAN_ACCESSORY("Motion Sensor");// Add the new Accessory
    new DEV_MotionSensor(motionSensorPin);// sensorPin or motionSensorPin ???
}

///////////////////////////////

void loop()
{
  homeSpan.poll();
}

///////////////////////////////
```

=> Compilation error: 'sensorPin' was not declared in this scope
Same issue if I use "sensorPin" instead of "motionSensorPin"
I don't see clearly what is wrong... ???

Greetings,
FiDelDworp

**Comment by HomeSpan** [comment]
"motionSensorPin" was just a placeholder.  You should use the actual sensor pin number (e.g. DEV_MotionSensor(9)).

**Reply by FidelDworp** [reply]
In the sketch above "sensorPin" is defined:
```cpp
  int sensorPin = 25;                            // pin number of the sensor
```

But if I use it instead of your "placeholder" I get the same error...

**Reply by FidelDworp** [reply]
OK! If I use "DEV_MotionSensor(25)":

  SPAN_ACCESSORY("Motion Sensor");// Add the new Accessory
    new DEV_MotionSensor(DEV_MotionSensor(25));// sensorPin or motionSensorPin ???

It compiled. Now testing...

**Reply by FidelDworp** [reply]
Wow! It works: The Motion sensor showed up in the serial monitor. And after 10 minutes I saw it appear in the Apple Home App on my Mac. Not yet on the iPhone... I managed to create an automation with it.

Thanks again for helping me!

(BTW: I am puzzled why I need to use "DEV_MotionSensor(25)" instead of "sensorPin" as defined in the structure... :-(

**Comment by HomeSpan** [comment]
This has to do with how C and C++ work.  Within the `struct DEV_MotionSensor` definition you created a variable *sensorPin*.   Since you defined it inside a structure, the variable is ONLY used inside any object created based on that structure.  The variable *sensorPin* is unknown outside of that structure.  When you call `new DEV_MotionSensor(sensorPin)`, the C compiler looks for a variable called *sensorPin* that has been defined within the `setup()` function and finds nothing.  It then checks to see if there is a global variable (defined outside of any function or structure) and still finds nothing.  Hence it fails to compile.

There are some good online C++ tutorials you may want to review - HomeSpan is fairly complicated and uses a lot of C++ techniques.

**Reply by FidelDworp** [reply]
Thanks for your generous help and tips Greg!
Indeed, now I understand. I will remove the local pin definition as it makes no sense.

I keep learning @ 70!
Greetz

**Comment by MrJelly17** [comment]
Can you share your full code please
